src/minecraft/kotlin/xyz/wagyourtail/unimined/minecraft/patch/fabric/AccessWidenerMerger.kt

```diff
@@ -0,0 +1,344 @@
+        transitive: Boolean
+    ) {
+        addOrMerge(
+            fieldAccess,
+            EntryTriple(owner, name, descriptor),
+            access,
+            FieldAccess.DEFAULT
+        )
+    }
+
+    private fun addOrMerge(
+        map: MutableMap<EntryTriple, Access>,
+        entry: EntryTriple,
+        access: AccessWidenerReader.AccessType,
+        defaultAccess: Access
+    ) {
+        map[entry] = applyAccess(access, map.getOrDefault(entry, defaultAccess), entry)
+    }
+
+    private fun applyAccess(
+        input: AccessWidenerReader.AccessType,
+        access: Access,
+        entryTriple: EntryTriple?
+    ): Access {
+        return when (input) {
+            AccessWidenerReader.AccessType.ACCESSIBLE -> {
+                makeClassAccessible(entryTriple)
+                access.makeAccessible()
+            }
+
+            AccessWidenerReader.AccessType.EXTENDABLE -> {
+                makeClassExtendable(entryTriple)
+                access.makeExtendable()
+            }
+
+            AccessWidenerReader.AccessType.MUTABLE -> access.makeMutable()
+            else -> throw java.lang.UnsupportedOperationException("Unknown access type:$input")
+        }
+    }
+
+    private fun makeClassAccessible(entryTriple: EntryTriple?) {
+        if (entryTriple == null) return
+        classAccess[entryTriple.owner] = applyAccess(
+            AccessWidenerReader.AccessType.ACCESSIBLE,
+            classAccess.getOrDefault(entryTriple.owner, ClassAccess.DEFAULT),
+            null
+        )
+    }
+
+    private fun makeClassExtendable(entryTriple: EntryTriple?) {
+        if (entryTriple == null) return
+        classAccess[entryTriple.owner] = applyAccess(
+            AccessWidenerReader.AccessType.EXTENDABLE,
+            classAccess.getOrDefault(entryTriple.owner, ClassAccess.DEFAULT),
+            null
+        )
+    }
+}
+
+class AccessWidenerBetterRemapper
+/**
+ * @param delegate      The visitor to forward the remapped information to.
+ * @param toNamespace   The namespace that the access widener will be remapped to.
+ */(
+    private val delegate: AccessWidenerVisitor,
+    private val mappingsProvider: MappingsProvider,
+    private val toNamespace: String,
+    private val mcProvider: MinecraftProvider<*, *>
+) : AccessWidenerVisitor {
+    private var remapper: TinyRemapper
+
+    init {
+        remapper = TinyRemapper.newRemapper()
+            .withMappings(
+                mappingsProvider.getMappingsProvider(
+                    EnvType.COMBINED,
+                    MappingNamespace.OFFICIAL to MappingNamespace.getNamespace(toNamespace),
+                    false
+                )
+            ).build()
+
+        remapper.readClassPathAsync(*mcProvider.mcLibraries.resolve().map { it.toPath() }.toTypedArray())
+        remapper.readClassPathAsync(mcProvider.getMinecraftWithMapping(EnvType.COMBINED, MappingNamespace.OFFICIAL, MappingNamespace.getNamespace(toNamespace)))
```

Comment:
this won't work iirc... official can't be remapped directly to yarn

Reply:
just use MappingNamespace.getNamespace(toNamespace) for both args

Reply:
also those are backwards as is... the fallback is the to namespace there

Reply:
why do we need a remapper in the first place?

Reply:
We need a remapper in case the input AW isn't of same namespace as the resulting merged AW.

Why will it not work from official to yarn?

Reply:
I'd prefer to just throw if a user tries to merge aw's in different namespaces.
as for the getMinecraftWithMappings, it's half the arg order https://github.com/WagYourOrg/unimined/blob/8da98da8078afaa12b1baa3e8619964192a97293/src/minecraft/kotlin/xyz/wagyourtail/unimined/minecraft/MinecraftProviderImpl.kt#L242 and half that remapping from official to yarn directly should throw on
https://github.com/WagYourOrg/unimined/blob/8da98da8078afaa12b1baa3e8619964192a97293/src/api/kotlin/xyz/wagyourtail/unimined/api/mappings/MappingNamespace.kt#L27
I think there's also some other issues with the way you plan to remap... but you can probably discover those if you create a few tests.